File: yodnsconf.rb
###
# Program:: http://www.docunext.com/wiki/Sinatra
# Component:: yodnsconf.rb
# Copyright:: Savonix Corporation
# Author:: Albert L. Lash, IV
# License:: Gnu Affero Public License version 3
# http://www.gnu.org/licenses
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with this program; if not, see http://www.gnu.org/licenses
# or write to the Free Software Foundation, Inc., 51 Franklin Street,
# Fifth Floor, Boston, MA 02110-1301 USA
##
require 'sinatra/base'
require 'rack/contrib/config'
require 'rack-rewrite'
require 'builder'
require 'sass'
require 'yajl'
require 'xml/xslt'
require File.dirname(File.dirname(__FILE__)) + '/Rack-XSLView/lib/rack-xslview' if ENV['RACK_ENV'] == 'development'
require 'rack-xslview' unless ENV['RACK_ENV'] == 'development'
require 'rack/cache'
require 'sinatra/xslview'
require 'sinatra/simplerdiscount'

require File.dirname(File.dirname(__FILE__)) + '/sinatra-stuff/sinatra-bundles/lib/sinatra/bundles' if ENV['RACK_ENV'] == 'development'
require 'sinatra/bundles' unless ENV['RACK_ENV'] == 'development'

require 'rexml/document'
require 'memcache'
require 'zonefile'
require 'net/dns/resolver'
require File.dirname(File.dirname(__FILE__)) + '/svxbox/lib/svxbox' if ENV['RACK_ENV'] == 'development'
require 'svxbox' unless ENV['RACK_ENV'] == 'development'

module Sinatra
  module ModBox
    include SvxBox::Sinatricus
    include SvxBox::MarkupGuppy
  end
  helpers ModBox
end

# The container for the Yodnsconf application
module Yodnsconf

  class << self
    attr_accessor(:conf, :memcdb)
  end

  # Create the app which will run
  def self.new(conf)
    self.conf = conf
    if self.conf[:ccf]
      puts self.conf[:ccf]
      if File.exists?(self.conf[:ccf])
        puts "reading custom conf"
        json = File.new(self.conf[:ccf], 'r')
        parser = Yajl::Parser.new(:symbolize_keys => true)
        customconf = parser.parse(json)
        self.conf.merge!(customconf)
      end
    end
    if self.conf[:memc_srv]
      self.memcdb = MemCache.new self.conf[:memc_srv], :namespace => self.conf[:memc_ns]
    end
    Main
  end
  class ZonefileNotFound < Sinatra::NotFound
    attr_reader :name

    def initialize(name)
      @name = name
    end
  end

  # The sub-classed Sinatra application
  class Main < Sinatra::Base
    set :environment => ENV['RACK_ENV']
    set :js => 's/js'
    set :css => 's/css'
    register Sinatra::Bundles
    puts ENV['RACK_ENV']

    configure do
      set :static => true, :public => 'web'
      set :xslviews => 'views/xsl/'
      set :uripfx => Proc.new { Yodnsconf.conf[:uripfx] }
      set :started_at => Time.now.to_i
      set :zonegroups => Proc.new { Yajl::Parser.parse(File.new(Yodnsconf.conf[:zonegroups], 'r'), :symbolize_keys => true) }

      # Set request.env with application mount path
      use Rack::Config do |env|
        env['analytics_key'] = Yodnsconf.conf[:analytics_key] if Yodnsconf.conf[:analytics_key]
        env['RACK_ENV'] = ENV['RACK_ENV']
        env['path_prefix'] = uripfx
        env['link_prefix'] = uripfx
        env['TS'] = Time.now.to_i
      end

      use Rack::Rewrite do
        rewrite Yodnsconf.conf[:uripfx]+'zones', '/s/xhtml/zones.html'
        rewrite Yodnsconf.conf[:uripfx]+'welcome', '/s/xhtml/welcome.html'
      end

      myxslfile = File.open('views/xsl/html_main.xsl') { |f| f.read }
      myxsl = XML::XSLT.new()
      myxsl.xsl = REXML::Document.new myxslfile
      set :xsl, myxsl
      set :xslfile, myxslfile
      set :noxsl, ['/raw/', '/s/img/', '/s/js/']
      set :passenv, ['PATH_INFO', 'RACK_MOUNT_PATH', 'RACK_ENV','link_prefix','path_prefix','analytics_key','TS']

    end
    configure :development do
      set :logging => false, :reload_templates => true
    end
    configure :demo do
      set :logging => true, :reload_templates => false
    end

    if ENV['RACK_ENV'] == 'production' && Yodnsconf.conf[:cache_base]
      use Rack::Cache,
        :verbose     => true,
        :metastore   => Yodnsconf.conf[:cache_base]+'/meta',
        :entitystore => Yodnsconf.conf[:cache_base]+'/body'
    end

    # Use Rack-XSLView
    use Rack::XSLView,
      :myxsl => xsl,
      :noxsl => noxsl,
      :passenv => passenv,
      :xslfile => xslfile,
      :reload => ENV['RACK_ENV'] == 'development' ? true : false,
      :content_type => 'application/xhtml+xml'

      set(:bundle_cache_time,0)
      set(:warm_bundle_cache,0)
      set(:cache_bundles,0)
 
      stylesheet_bundle(:all, %w(droppy yui-reset-min thickbox))
      javascript_bundle(:all, %w(jquery/jquery-1.4.2.min jquery/plugins/jquery.quicksearch jquery/plugins/jquery.droppy jquery/plugins/jquery.history-201006 jquery/plugins/jquery.pagination))

    # Sinatra Helper Gems
    helpers Sinatra::XSLView
    helpers Sinatra::ModBox
    helpers Sinatra::SimpleRDiscount

    helpers do
      def link_prefix
        "#{Yodnsconf.conf[:uripfx]}"
      end
      def rootxml
        '<root />'
      end
      # These should be different based upon development vs. production
      def get_servers(domain=nil)
        servers = []
        server = Hash.new
        server[:id] = 1
        server[:server] = 'Franklin'
        server[:host_name] = 'mx2.example.com'
        server[:modified] = Time.now.to_i
        servers << server
        return servers
      end
      def get_domains(domain_group=nil)
        # TODO caching
        zone = Yodnsconf.conf[:zonelist]
        zf = Zonefile.from_file(zone)
        idx_json = []
        zf.txt.each do |res|
          unless res[:ttl][0].to_i == 2
            if domain_group.nil?
              idx_json << res[:text]
            else
              idx_json << res[:text] if res[:ttl][1].to_i == domain_group
            end
          end
        end
        return idx_json
      end

      def open_zonefile(zone)
        zonefile = "data/zones/#{zone}.zone"
        if File.exists?(zonefile)
          return Zonefile.from_file(zonefile)
        else
          raise ZonefileNotFound.new(zone)
        end
      end

      def check_rr_type(type)
        Zonefile::RECORDS.include?(type)
      end

      def get_public_ns(domain)
        res = Net::DNS::Resolver.new(:nameservers => "8.8.4.4")
        packet = res.query(domain, Net::DNS::NS)
        ns = packet.answer.map do |rr|
          rr.nsdname
        end
        return ns
      end
    end

    before do
      cache_control :'no-store'
    end
    get '/' do
      mredirect 'welcome'
    end
    get %r{/(host|record|zone-group|service|redirect|ip)s$} do
      xslview rootxml, params[:captures].first.gsub('-','_') + 's.xsl', { 'link_prefix' => link_prefix }
    end
    get %r{/(host|zone-group|service|redirect|ip)-edit} do
      xslview rootxml, params[:captures].first.gsub('-','_') + '_edit.xsl', { 'link_prefix' => link_prefix }
    end

    get '/zone/:zone/:type' do
      zf = open_zonefile(params[:zone].to_s)
      type = params[:type].to_s
      begin
        check_rr_type(type)
        rs = zf.send type.to_sym
        rsarr = []
        if rs
          rs[0][:type] = type unless rs[0].nil?
          rsarr << rs[0] unless rs[0].nil?
        end
        if type=='soa'
          rsarr = rs
        end
        "<pre>#{rsarr.to_s}</pre>"
      rescue NoMethodError
        '<span class="errblk">Not a valid record type.</span>'
      end
    end
    
    get '/raw/yd-zones/' do
      xslview rootxml, 'zones.xsl'
    end
    get '/raw/json/yd-zone-groups' do
      content_type :json
      zg = File.open('data/json/zone_groups.json') { |f| f.read }
      zg.to_s
    end
    get '/raw/json/zones' do
      content_type :json
      get_domains.to_json
    end
    get '/raw/json/zones/:group' do
      content_type :json
      get_domains.to_json
    end
    get '/raw/json/records/:type/:zone' do
      content_type :json
      zone = params[:zone].to_s
      zf = open_zonefile(params[:zone].to_s)
      type = params[:type].to_s
      rs = zf.send type.to_sym
      rsarr = []
      if rs
        rs[0][:type] = type unless rs[0].nil?
        rsarr << rs[0] unless rs[0].nil?
      end
      if type=='soa'
        rsarr = rs
      end
      rsarr.to_json
    end
    get '/raw/json/public-ns/:zone' do
      content_type :json
      zone = params[:zone].to_s
      ns = get_public_ns(zone)
      ns.to_json
    end
    not_found do
      cache_control :'no-store', :max_age => 0
      %(<div class="block"><div class="hd"><h2>Error</h2></div><div class="bd">This is nowhere to be found. <a href="#{Yodnsconf.conf[:uripfx]}">Start over?</a></div></div>)
    end

    get '/s/css/stylesheet.css' do
      cache_control :public, :max_age => 600
      content_type 'text/css', :charset => 'utf-8'
      sass :yodnsconf
    end
  end

end

File: web/s/js/yodnsconf.js
/*
###
Program: YoDNSConf
Component: yodnsconf.js
Copyright: Savonix Corporation
Author: Albert L. Lash, IV
License: Gnu Affero Public License version 3
http://www.gnu.org/licenses

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program; if not, see http://www.gnu.org/licenses
or write to the Free Software Foundation, Inc., 51 Franklin Street,
Fifth Floor, Boston, MA 02110-1301 USA
##
*/

function create_row(zone,myr) {
  var newtextnode = document.createTextNode(zone);
  mytd.appendChild(newtextnode);
  return mytr;
}

function build_list(data,options) {
  
}
function build_table(table,data,options,key) {
    index = $(table).data('index');
    zones = data;
    zoneview = $(table).data(key).slice(index,index + 10);

    var mytb = document.createElement("tbody");
    var mytr = document.createElement("tr");
    var mytd1 = document.createElement("td");
    mytd1.className = 'primary-key';
    var mytd2 = document.createElement("td");
    for(var i=0,max=options.length;i<max;i++){
      var mysp1 = document.createElement("span");
      var mytn1 = document.createTextNode(options[i]);
      mysp1.appendChild(mytn1);
      mytd2.appendChild(mysp1);
    }
    
    $.each(zoneview, function(i,zone) {
        myr = mytr.cloneNode(true);
        var mytdd1 = mytd1.cloneNode(true);
        var mytdd2 = mytd2.cloneNode(true);
        myr.appendChild(mytdd1);
        myr.appendChild(mytdd2);
        var newtextnode = document.createTextNode(zone);
        mytdd1.appendChild(newtextnode);
        mytb.appendChild(myr);
    });
    $('tbody',table).replaceWith(mytb);
    $('.primary-key',table).click(function(){
        thezone = $(this).text();
        load_zone(thezone);
    });
    $(table).data('index', index + 5);
    $('#origin').quicksearch('tbody tr',table);
}
function load_zones() {
  $('#content').load(app_prefix + 'raw/xhtml/zones.html', function() {
    $.getJSON(app_prefix + 'raw/json/zones',function(data) {
        $('#myzones').data('zones', data);
        $('#myzones').data('index', 0);
        var options = ["Edit", "Delete", "Clone"];
        build_table('#myzones',data,options,'zones');
    });
  });
}
function load_redirects() {
  $('#content').load(app_prefix + 'raw/xhtml/redirects.html', function() {
    $.getJSON(app_prefix + 'raw/json/redirects',function(data) {
        $('#myzones').data('zones', data);
        $('#myzones').data('index', 0);
        var options = ["Edit", "Delete", "Clone"];
        build_table('#myzones',data,options,'zones');
    });
  });
}
function load_zone(thezone) {
  document.title = thezone;
  $.history.load(thezone);
  $('#content').load(app_prefix + 'raw/xhtml/zone_edit.html', function() {
      $('input:text[name="origin[]"]').val(thezone);
      $.getJSON(app_prefix + 'raw/json/records/soa/'+thezone,function(data) {
          for (var key in data) {
            $('input:text[name="'+key+'"]').val(data[key]);
          }
      });
      $.getJSON(app_prefix + 'raw/json/public-ns/'+thezone,function(data) {
          for (var i=0;i<data.length;i=i+1) {
            var ns = data[i] + ", ";
            $('#public-resolvers .data').append(ns);
          }
      });
      $.getJSON(app_prefix + 'raw/json/yd-zone-groups',function(data) {
          var opt = $('#zone_group_id').find('option').clone();
          for (var i=0;i<=data.length;i=i+1) {
            var zg = data[i];
            opt.clone().val(zg.id).text(zg.label).appendTo($('#zone_group_id'));
          }
      });
  });
}
function load_records(thezone) {
  
}

$(document).ready(function() {
  $('#nav').droppy();
  if($('#myzones').length > 0) {
    $.history.init(function(url) {
      if(url == "") {
        load_zones();
      } else {
        load_zone(url);
      }
    });
  }
  if($('#myredirects').length > 0) {
    $.history.init(function(url) {
      if(url == "") {
      } else {
      }
    });
  }
});





/* DEPRECATED */


function delete_redirect(http_host,pos) {
    if(confirm('Are you sure?')){
      $.post(app_prefix + "x-redirect-delete&http_host="+http_host,
      {
        'http_host': http_host
      },
      function (data){
        $("#h_"+pos).remove();
      });
    }
}

function delete_host(id) {
  if(confirm('Are you sure?')){
    $.post(app_prefix + "x-host-delete&host_id="+id,
    {
      'host_id': id
    },
    function (data){
      $("#h_"+id).remove();
    });
  }
}

function delete_ip(ip_id) {
    if(confirm('Are you sure?')){
    $.post(app_prefix + "x-ydc-ip-addr-delete&ip_id="+ip_id,
    {
        'ip_id': ip_id
    },
    function (data){
      $("#ip_"+ip_id).remove();
    });
    }
}

function delete_record(record_id,zone) {
  if(confirm('Are you sure?')){
    $.post(app_prefix + "x-record-delete&record_id="+record_id,
    {
      // zone is needed to update serial
      'record_id': record_id,
      'zone': zone
    },
    function (data){
      $("#"+record_id).remove();
    });
  }
}
function delete_hostgroup(id) {
  if(confirm('Are you sure?')){
    $.post(app_prefix + "x-hostgroup-delete&hostgroup_id="+id,
    {
      'hostgroup_id': id
    },
    function (data){
      $("#hg_"+id).remove();
    });
  }
}
// zone delete / delete_zone
function zd(zone) {
  if(confirm('Delete zone id '+zone+'? You can deactivate it instead.')){
    $.post(app_prefix + "x-zone-delete&zone="+zone,
    {
      'zone': zone
    },
    function (data){
      $("#"+zone).remove();
    });
    // TODO - FIXME - Use jQuery AJAX.
    $("#"+zone).remove();
  }
  return false;
}
function delete_zone_group(zone_group_id) {
  if(confirm('Are you sure you want to ?')){
    $.post(app_prefix + "x-zone-group-delete&zg_id="+zone_group_id,
    {
      'zone_group_id': zone_group_id
    },
    function (data){
      $("#zg_"+zone_group_id).remove();
    });
    // TODO - FIXME - Use jQuery AJAX.
    $("#zg_"+zone_group_id).remove();
  }
  return false;
}


function delete_service(id) {
    if(confirm('Are you sure?')){
    $.post(app_prefix + "x-service-delete&service_id="+id,
    {
      'service_id': id
    },
    function (data){
      $("#s_"+id).remove();
    });
    }
}

function select_all_boxes(checkbox_name) {
  $(":checkbox[name='"+checkbox_name+"[]']").attr("checked","checked");
}
